### DISCUSSION TITLE
Tip for anyone using pickle and getting `KeyError('__pydantic_fields_set__')`

### DISCUSSION THREAD

**Discussion by danstewart** [pydantic/pydantic#6766] [Pydantic V2]
We have just upgraded to pydantic V2 but unfortunately we have some pickled pydantic V1 instances that we still need to fetch out.

Due to some renaming inside pydantic this now fails with:
```
KeyError('__pydantic_fields_set__')
```

There is a workaround for this though if you are subclassing `BaseModel`, just add this:
```python3
def __setstate__(self, state: dict[Any, Any]) -> None:
    """
    Hack to allow unpickling models stored from pydantic V1
    """
    state.setdefault("__pydantic_extra__", {})
    state.setdefault("__pydantic_private__", {})

    if "__pydantic_fields_set__" not in state:
        state["__pydantic_fields_set__"] = state.get("__fields_set__")

    super().__setstate__(state)
```

---

For any new code it's probably a better idea to pickle the dict of the model (`model_instance.model_dump()`) and convert it back into the model yourself.

**Comment by bkappes** [comment]
I run into an Attribute Error (`AttributeError: 'super' object has no attribute '__setstate__'`). The following has a similar goal as above, respect the V2 convention but fallback to V1 naming, and got around the AttributeError.

```python
def __setstate__(self, state: dict[Any, Any]) -> None:
        _object_setattr(self, '__pydantic_fields_set__',
                        state.get('__pydantic_fields_set__',
                                  state.get("__fields_set__", {})))
        _object_setattr(self, '__pydantic_extra__',
                        state.get('__pydantic_extra__', {}))
        _object_setattr(self, '__pydantic_private__',
                        state.get('__pydantic_private__', {}))
        _object_setattr(self, '__dict__', state['__dict__'])
```

**Reply by kaine-bruce-dmt** [reply]
Thanks! This helped me out :)

**Reply by Lmex89vexi** [reply]
hello, 

  File "pydantic/main.py", line 417, in pydantic.main.BaseModel.__setstate__
    object_setattr(self, '__fields_set__', state['__fields_set__'])
    └ <slot wrapper '__setattr__' of 'object' objects>
is this is the same problem?
